### PR TITLE
Refactored usage of storage traits

### PR DESCRIPTION
**Pull request by xgreenx**
Updated the crate to use https://github.com/FuelLabs/fuel-storage/pull/15

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -11,7 +11,7 @@ description = "Fuel Merkle tree libraries."
 
 [dependencies]
 digest = { version = "0.10", default-features = false }
-fuel-storage = "0.2"
+fuel-storage = { git = "https://github.com/FuelLabs/fuel-storage", branch = "feature/split-storage-trait" }
```
Comment (vlopes11) [inline Cargo.toml:14]: Consider publishing `fuel-storage` before proceeding with this merge

**Review by vlopes11 — CHANGES_REQUESTED** [review]

File: src/binary/merkle_tree.rs [hunk 2]
```diff
@@ -35,9 +35,19 @@ pub struct MerkleTree<StorageType> {
     leaves_count: u64,
 }
 
+/// Flat array of merkle's tree nodes.
+pub struct NodesArray;
```
Comment (Voxelot) [inline src/binary/merkle_tree.rs:39]: Maybe it would be more appropriate to call this a `NodesIndex`, since the purpose is to correlate u64's with Node's? It seems a little presumptuous to assume it's an array.

Comment (xgreenx) [reply src/binary/merkle_tree.rs:39]: Yeah, it is wrong naming, `MerkleTree` works with it as with array, but the storage can be anything=)
`NodesStorage` or `NodesTable` should be better. The comment will describe that the key is the `u64` index.

**Review by Voxelot — APPROVED** [review]
lgtm, waiting on @bvrooman to also approve

**Review by bvrooman — COMMENTED** [review]
I have some small changes and I'd like to see at this point, and to discuss the TODO comment. Overall, I think the changes are in good shape.

File: src/sum/merkle_tree.rs [hunk 3]
```diff
@@ -13,14 +13,49 @@ pub enum MerkleTreeError {
...
+/// fee: a.fee + b.fee
+/// data: node_sum(a.fee, a.data, b.fee, b.data)
+///
+/// where `node_sum` is defined as the hash function described in the data pair description above.
 pub struct MerkleTree<StorageType> {
     storage: StorageType,
     head: Option<Box<Subtree<Node>>>,
 }
 
+/// The table of the binary merkle's sum tree nodes. The storage key is the `Byte32` hash
```
Comment (bvrooman) [inline src/sum/merkle_tree.rs:44]:
```suggestion
/// The table of the Binary Merkle Sum Tree's nodes. The storage key is the `Byte32` hash
```

File: src/sparse/merkle_tree.rs [hunk 4]
```diff
@@ -30,9 +30,24 @@ pub struct MerkleTree<StorageType> {
     storage: StorageType,
 }
 
+/// The table of the sparse merkle's tree nodes. [`MerkleTree`] works with it as a sparse merkle
```
Comment (bvrooman) [inline src/sparse/merkle_tree.rs:33]:
```suggestion
/// The table of the Sparse Merkle tree's nodes. [`MerkleTree`] works with it as a sparse merkle
```

File: src/common/storage_map.rs [hunk 5]
```diff
@@ -1,57 +1,71 @@
...
+{
+    fn insert(
+        &mut self,
+        key: &Type::Key,
+        value: &Type::SetValue,
+    ) -> Result<Option<Type::GetValue>, Self::Error> {
+        let previous = self.map.remove(key);
+
+        self.map.insert(key.clone(), value.clone().into());
+        Ok(previous)
```
Comment (bvrooman) [inline src/common/storage_map.rs:60]: It looks like you are changing the behaviour of the insert to return the previous value at `key`. At the time of writing the original `insert`, there was no standard definition, AFAIK. Is returning the previous value the proper idiom?

Comment (xgreenx) [reply src/common/storage_map.rs:60]: Yes, it is how [`HashMap` works](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.insert) in Rust. Insert value `X` and return value `Some(X)` looks strange=D

File: src/sparse/merkle_tree.rs [hunk 6]
```diff
@@ -30,9 +30,24 @@ pub struct MerkleTree<StorageType> {
...
+/// tree, where the storage key is `Bytes32` and the value is the [`Buffer`](crate::sparse::Buffer)
+/// (raw presentation of the [`Node`](crate::sparse::Node)).
+pub struct NodesTable;
+
+impl Mappable for NodesTable {
+    /// The 32 bytes unique key of the merkle node.
+    type Key = Bytes32;
+    /// The merkle node data with information to iterate over the tree.
+    // TODO: Use directly [`Node`](crate::sparse::Node) because it will not add overhead.
```
[the remark this answers is not in the document]
Comment (xgreenx) [reply src/sparse/merkle_tree.rs:42]: It is more about consistent usage of `Node` in the binary merkle tree, sparse merkle tree, and binary merkle sum tree=)

Usage of `Buffer` in one of them and `Node` in two others look not logical. To use `Buffer`, the developer still needs to convert it into `Node`(and do the reverse operation to put it into the storage).

The `sparse::Node` can store it as a raw array under the hood. We don't need to use `serde`, the binary format is suitable=) The problem can be that `Node` doesn't implement serialization and deserialization(trait) required by Postgres or MySql. But it can be solved in several ways:
- On the storage side, `Node` can be wrapped into a new struct, and traits can be implemented for this struct.
- We can implement those traits in `fuel-merkle`, but we need to know which traits to implement(it breaks abstraction).
- We can implement `AsRef<[u8]>`, `From<Buffer>`, `From<Node>` traits, and on the storage side, we still will work with slice/array and can use binary format. (I like that approach, but we will have a problem during deserialization because we need to use `Buffer` and after `into`)

In this case, we don't have any performance impact. And code looks consistent, and it is easy to use.

[the remark this answers is not in the document]
Comment (bvrooman) [reply src/sparse/merkle_tree.rs:42]: Consistency is good! I agree that we should have a consistent pattern across all Merkle tree storage types. Actually, I was considering refactoring the Binary and Binary Sum tree to use a similar pattern as the Sparse tree, but I didn't come to a conclusion on that yet. This discussion is all part of that decision!

On reason I wanted to hide the use of `Node` (Sparse, Binary, or Binary Sum) from the `Storage` backings is that `Node` is an internal concept. It should only be used inside the tree implementation itself. None of the public methods on the trees actually take in or return a `Node`: they all operate on binary arrays (`&[u8]` or `Bytes32`). `Buffer` is simply a type alias for `[u8; 69]`. 

But as you suggest, we can still implement traits on top of a publicly exposed struct. We can even call that public struct `Node` too, since that is intuitive. 

Ultimately, I agree with the goals you are trying to achieve, and I think we should investigate further how to achieve them. At this moment, I think further discussion is outside the scope of this PR and I don't want to block it. We can continue this discussion at a later point!

**Review by bvrooman — APPROVED** [review]
Thanks for applying the suggestions!

**Review by Voxelot — APPROVED** [review]

**Review by vlopes11 — APPROVED** [review]

File: src/sparse/merkle_tree.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Voxelot) [reply src/sparse/merkle_tree.rs:42]: @bvrooman could you summarize this into a backlog issue so we don't lose track of this concern?

[the remark this answers is not in the document]
Comment (bvrooman) [reply src/sparse/merkle_tree.rs:42]: Created an issue here: https://github.com/FuelLabs/fuel-merkle/issues/113
